src/BrazeBannerComponent.ts

```diff
@@ -20,12 +20,16 @@ import {
     StyleableBannerWithLink,
 } from './StyleableBannerWithLink';
 import { COMPONENT_NAME as BANNER_NEWSLETTER_NAME, BannerNewsletter } from './BannerNewsletter';
+import {
+    COMPONENT_NAME as STYLEABLE_BANNER_NEWSLETTER_NAME,
+    StyleableBannerNewsletter,
+} from './StyleableBannerNewsletter';
 
 export type CommonBannerComponentProps = {
     componentName: string;
     brazeMessageProps: BrazeMessageProps;
-    subscribeToNewsletter: NewsletterSubscribeCallback;
-    fetchEmail: FetchEmail;
+    subscribeToNewsletter?: NewsletterSubscribeCallback;
+    fetchEmail?: FetchEmail;
```

Comment:
I don't think we should make these optional. In the current model there's a single "factory" component  - `BrazeBannerComponent` through which all other components are rendered. For example, DCR only knows about this component and not the different banner components underneath. So from this component's perspective I think these should _always_ be required, since from a types perspective we can't know for sure that they won't be needed. This may change when components are moved to DCR, but for now what do you think about putting this back to how it was? If these were made optional, they could be removed from the DCR call to `BrazeBannerComponent` and TS would be happy, but it would break the functionality around subscribing to newsletters or subscribing to reminders.

Reply:
For me, this is one of those situations where TS is getting in the way of the real code. I'll try to explain.

DCR passes through two functions to its "factory" component, which we then **need to check in real code** to make sure they have indeed been passed through before rendering the affected component or banner.

- For `StyleableBannerWithLink` this is the `fetchEmail` function used to set up a reminder when the user clicks on the "Remind me in ..." button. If for any reason the function is missing then that button won't render - but the rest of the banner will, as banners with links can be about anything, not just a contributions request.

- For `StyleableBannerNewsletter` the only CTA is the "Subscribe to newsletter" button. Without the `subscribeToNewsletter` function there's no point showing the user the banner (bad UX) so when it's absent, the banner returns `null` rather than some HTML.

Newsletter banners don't need to know about `fetchEmail`, while link banners don't need to know about `subscribeToNewsletters` (the App banner does its own thing and doesn't care about either).

The current TS markup says that both of those functions are required by all banner components. That is a lie (and, in my view, bad documentation). It's up to the components to check in **real code** that the functions they require have indeed been passed to them by DCR - which, as code living outside of this repository must be treated as unreliable (and a security risk - anyone can change the relevant DCR code without our knowledge). I've added these checks to the components as part of this PR to mitigate the risk of the components breaking the main site when user tries to invoke a non-existent function.

Happy to chat about this IRL if you disagree.

Reply:
Hey thanks for writing this up! Yep totally understood that there may be bits that are sub-optimal at the moment but I don't think changing the interface to the _outer_ factory component is the way to address those things. You're right that the individual inner components don't always need all the props provided to the outer component but the consumer of the factory component doesn't/shouldn't have to be aware of that in the current setup I don't think. All the caller knows is, "I've got some props to pass from Braze which _might_ result in any of the underlying components being rendered". So it has to pass all the dependencies. I'm all for making sure the inner components' interfaces reflect reality. Regardless of how this conversation ends up perhaps this could be done in a separate PR (I don't think it's directly related to or depended on by the new banner component).

Reply:
Reverted.